thisdot/lets-chat-with
# Show proper chatter and chats data

# Overview

**NOTE**: The api.service file and all the mapping-template file changes can be ignored during review - those are auto-generated from the amplify-schema file changes.

Closes #44 
Updates the GraphQL schema and relevant components / queries so that when a user views their list of conferences, they see the correct data for the chatters and chats fields. 
Chatters now shows the total number of active attendees for that event
Chats is still the number of active chats the user has for that event, updated to show the correct data

# Screenshots for Mobile and Desktop views (if applicable)

## Before

![Screen Shot 2023-03-09 at 12 59 31 PM](https://user-images.githubusercontent.com/16214572/224127595-216cb0bc-4d0a-4d58-b6b3-6b38666240ab.png)

## After

![Screen Shot 2023-03-09 at 12 50 24 PM](https://user-images.githubusercontent.com/16214572/224127616-48554a2d-6be7-4594-b59b-32f85e159ba0.png)

# Details

## General

- Got the true attendee count
  - Updated the amplify schema for the Event type so the attendees connection matched on a keyName and the relevant field id, which now lets events actually see the attendees data
  - Updated the conference-card component (and the relevant models) so they have an attendeeCount field that pulls in the list of attendees for an event and shows that count for the chatters field
- Got the true chat count
    - In the amplify schema, had to update the Match type to allow for private read-only access so the ChatThread type could see that data
    - Then updated the chatThread resolver so it can filter the full list of chats for the event to just ones that the current attendee would have knowledge of

### Manual Testing

For testing, I needed two separate browser instances and it's easiest to verify if run locally (though this should work just as well on the deployed instance, just more data to sort through)

1. Run `ng serve`
2. In browser 1, sign in with a user and go to localhost:4200/conferences
3. If running locally, should see a non-0 value for chatters (the exact number depends on number of test users you created during initial setup) and 0 chats
4. Click into the event and connect with at least one other user that you have access to
5. In browser 2, log in with the user you matched with in the previous step and go to localhost:4200/conferences
6. Since this user hasn't agreed to the match yet, they should still see a non-0 value for chatters and 0 chats
7. Click into the event and connect with the user from browser 1
8. Both users should now see that the match has been established and be able to see the other user in their matches and chat sections
9. For either / both user(s), go back to the conferences list. Should now see that the chat count has been updated to 1.
10. Bonus step: log out from one of the users and log in with a different user (one that hasn't matched with anyone yet). If you now visit the conference list page, they should still see 0 chats.
